lib/SILOptimizer/Mandatory/MandatoryCombine.cpp

```diff
@@ -0,0 +1,222 @@
+///  instruction combines to be performed at the begining of both the Onone and
+///  also the performance pass pipelines, after the diagnostics passes have been
+///  run.  It is intended to be run before and to be independent of other
+///  transforms.
+///
+///  The intention of this pass is to be a place for mandatory peepholes that
+///  are not needed for diagnostics. Please put any such peepholes here instead
+///  of in the diagnostic passes.
+///
+//===----------------------------------------------------------------------===//
+
+#define DEBUG_TYPE "sil-mandatory-combiner"
+#include "../SILCombiner/SILCombiner.h"
+#include "swift/SIL/SILVisitor.h"
+#include "swift/SILOptimizer/PassManager/Passes.h"
+#include "swift/SILOptimizer/PassManager/Transforms.h"
+#include "swift/SILOptimizer/Utils/Local.h"
+#include "llvm/ADT/STLExtras.h"
+#include "llvm/Support/raw_ostream.h"
+#include <algorithm>
+
+using namespace swift;
+
+namespace {
+
+class MandatoryCombiner final
+    : public SILInstructionVisitor<MandatoryCombiner, SILInstruction *> {
+
+  using Worklist = SmallSILInstructionWorklist<256>;
+
+  /// The list of instructions remaining to visit, perhaps to combine.
+  Worklist worklist;
+
+  /// Whether any changes have been made.
+  bool madeChange;
+
+  /// The number of times that the worklist has been processed.
+  unsigned iteration;
+
+  InstModCallbacks instModCallbacks;
+  llvm::SmallVectorImpl<SILInstruction *> &createdInstructions;
+  llvm::SmallVector<SILInstruction *, 16> instructionsPendingDeletion;
+
+public:
+  MandatoryCombiner(
+      llvm::SmallVectorImpl<SILInstruction *> &createdInstructions)
+      : worklist("MC"), madeChange(false), iteration(0),
+        instModCallbacks(
+            [&](SILInstruction *instruction) {
+              worklist.erase(instruction);
+              instructionsPendingDeletion.push_back(instruction);
+            },
+            [&](SILInstruction *instruction) { worklist.add(instruction); }),
+        createdInstructions(createdInstructions){};
+
+  /// Base visitor that does not do anything.
+  SILInstruction *visitSILInstruction(SILInstruction *) { return nullptr; }
+
+  void addReachableCodeToWorklist(SILFunction &function) {
+    SmallBlotSetVector<SILBasicBlock *, 32> blockWorklist;
+    SmallBlotSetVector<SILBasicBlock *, 32> blocksVisited;
+    llvm::SmallVector<SILInstruction *, 128> instructions;
+
+    blockWorklist.insert(&*function.begin());
+    while (!blockWorklist.empty()) {
+      auto *block = blockWorklist.pop_back_val().getValueOr(nullptr);
+      if (block == nullptr) {
+        continue;
+      }
+
+      if (!blocksVisited.insert(block).second) {
+        continue;
+      }
+
+      for (auto iterator = block->begin(), end = block->end(); iterator != end;) {
+        auto *instruction = &*iterator;
+        instructions.push_back(instruction);
+        ++iterator;
+      }
+
+      std::for_each(block->succ_begin(), block->succ_end(), [&](SILBasicBlock *block) {
```

Comment:
I think we have an adapter for ranges for for_each. Can you use that instead? So this becomes:

```
for_each(block->getSuccessorBlocks(), [&](SILBasicBlock *block) {
   blockWorklist.insert(block);
});
```

Reply:
Ah, nice.  Fixed.